vllm/worker/multi_step_worker.py

```diff
@@ -0,0 +1,153 @@
+                assert model_input.outputs[-1].sampled_token_ids is not None
+
+        if self.do_metadata_broadcast:
+            broadcast_data = worker_input.as_broadcastable_tensor_dict()
+            broadcast_data.update(model_input.as_broadcastable_tensor_dict())
+            broadcast_tensor_dict(broadcast_data, src=0)
+
+        return model_input, worker_input
+
+    def prepare_input(
+        self,
+        execute_model_req: Optional[ExecuteModelRequest] = None,
+    ) -> Optional[Tuple[MutableModelInputForGPUWithMultiStepMetadata,
+                        WorkerInput]]:
+        """
+        Depending on the current state of the request and multi step worker,
+        this method may skip the normal _prepare_model_input and
+        _prepare_worker_input methods and instead used cached values.
+        """
+        if self.is_driver_worker:
+            if execute_model_req is None:
+                if self.do_metadata_broadcast:
+                    # This signals that there's no more requests to process for
+                    # now. All workers are running infinite loop with
+                    # broadcast_tensor_dict, and it stops the loop when the
+                    # driver broadcasts an empty input. Send an empty input to
+                    # notify all other workers to stop their execution loop.
+                    broadcast_tensor_dict({}, src=0)
+                return None
+
+            virtual_engine = execute_model_req.virtual_engine
+            model_input, worker_input = self._get_driver_input_and_broadcast(
+                execute_model_req)
+            assert isinstance(model_input,
+                              MutableModelInputForGPUWithMultiStepMetadata)
+            if execute_model_req.is_first_multi_step:
+                # cache the worker input and model input for the next steps
+                self.multi_step_states[virtual_engine] = MultiStepState(
+                    worker_input=worker_input, model_input=model_input)
+        # if TP workers
+        else:
+            broadcast_data = self._get_worker_input_from_broadcast()
+            # if the driver has sent an empty input, we should stop the worker
+            # loop
+            if broadcast_data is None:
+                return None
+            model_input, worker_input = broadcast_data
+            assert isinstance(model_input,
+                              MutableModelInputForGPUWithMultiStepMetadata)
+            virtual_engine = worker_input.virtual_engine
+            if model_input.is_first_multi_step:
+                pass
+                # cache the worker input and model input for the next steps
+                # TODO(will) see below
+            else:
+                # TODO(will) possible to also use the cached worker input and
+                # model input this can be done if we want to optimize the
+                # broadcast to only send the last sampled token ids for
+                # non-first multi steps
```

Comment:
Can you elaborate more and give an example on this part? I don't fully understand.

Reply:
Updated the comment for clarity. The idea is essentially the delta optimization for model_inputs. Where the TP workers can cache the model input states and we only broadcast the delta need for the next step (sampled_token_ids from the previous step)

Reply:
[9825ace](https://github.com/vllm-project/vllm/pull/7000/commits/9825ace98a746c6dddee347f9ff39e829f7ff415)